src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketInput.cs

```diff
@@ -45,6 +45,19 @@ public SocketInput(MemoryPool memory, IThreadPool threadPool, IBufferSizeControl
 
         public bool IsCompleted => ReferenceEquals(_awaitableState, _awaitableIsCompleted);
 
+        public bool CheckFinOrThrow()
```

Comment:
Ooh I see what you mean, tri-state so we avoid races between `RemoteIntakeFin` and `_awaitableError` being inconsistent with each other, right?